File: apps/tabtin-android/app/src/main/java/com/tabtin/mobile/data/im/ImConversationTitlePolicy.kt
package com.tabtin.mobile.data.im

import java.util.UUID

/** 客户端会话标题策略：DM 显示对端成员名，不暴露传输层的系统占位名。 */
internal object ImConversationTitlePolicy {
    private val providerFallbacks = setOf(
        "tabtin private conversation",
        "private conversation",
    )

    fun resolve(
        conversationName: String,
        isDirectMessage: Boolean,
        peerDisplayName: String?,
        directMessageFallback: String,
        conversationFallback: String,
    ): String {
        val name = conversationName.trim()
        if (!isDirectMessage) return name.ifEmpty { conversationFallback }

        peerDisplayName?.trim()?.takeIf(::isReadableDirectMessageName)?.let { return it }
        if (isReadableDirectMessageName(name) && name.lowercase() !in providerFallbacks) return name
        return directMessageFallback
    }

    private fun isReadableDirectMessageName(value: String): Boolean =
        value.isNotEmpty() && runCatching { UUID.fromString(value) }.isFailure
}

File: apps/tabtin-ios/Tabtin/Core/IM/IMConversationTitlePolicy.swift
import Foundation

/// 会话标题只消费产品语义，不把传输层的系统占位名暴露给用户。
enum IMConversationTitlePolicy {
    private static let providerFallbacks: Set<String> = [
        "tabtin private conversation",
        "private conversation",
    ]

    static func resolve(
        conversationName: String,
        isDirectMessage: Bool,
        peerDisplayName: String?,
        directMessageFallback: String = "私信",
        conversationFallback: String = "会话"
    ) -> String {
        let name = normalized(conversationName)
        guard isDirectMessage else {
            return name.isEmpty ? conversationFallback : name
        }

        let peer = normalized(peerDisplayName ?? "")
        if isReadableDirectMessageName(peer) { return peer }
        if !name.isEmpty,
           isReadableDirectMessageName(name),
           !providerFallbacks.contains(name.lowercased()) {
            return name
        }
        return directMessageFallback
    }

    private static func normalized(_ value: String) -> String {
        value.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func isReadableDirectMessageName(_ value: String) -> Bool {
        !value.isEmpty && UUID(uuidString: value) == nil
    }
}
